src/Common/src/Interop/Unix/libcurl/Interop.SafeCurlHandle.cs

```diff
@@ -56,11 +96,175 @@ public static void DisposeAndClearHandle(ref SafeCurlMultiHandle curlHandle)
                 }
             }
 
+            internal void PollFds(List<libc.pollfd> readyFds)
+            {
+                int count;
+                libc.pollfd[] pollFds;
+
+                readyFds.Clear();
+
+                lock (this)
+                {
+                    // TODO: Avoid the allocation when count is in 100s
+                    count = _fdSet.Count + 1;
+                    pollFds = new libc.pollfd[count];
+
+                    // Always include special fd in the poll set
+                    pollFds[0].fd = _specialFds[libc.ReadEndOfPipe];
+                    pollFds[0].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                    int i = 1;
+                    foreach (int fd in _fdSet)
+                    {
+                        pollFds[i].fd = fd;
+                        pollFds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                        i++;
+                    }
+                }
+
+                unsafe
+                {
+                    fixed (libc.pollfd* fds = pollFds)
+                    {
+                        int numFds = libc.poll(fds, (uint)count, -1);
+                        lock (this)
+                        {
+                            if (0 == _fdCount)
+                            {
+                                return;
+                            }
+                        }
+                        if (numFds <= 0)
```

Comment:
You probably want to move this if check up to before the lock, right?  That way we'll get the assert for the error condition regardless of whether _fdCount is 0?